hail/src/main/scala/is/hail/expr/ir/functions/LocusFunctions.scala

```diff
@@ -52,43 +48,45 @@ object LocusFunctions extends RegistryFunctions {
     registerLocusCode("inXNonPar")
     registerLocusCode("inYPar")
 
-    registerCode("min_rep", tv("T", "locus"), TArray(TString()), TStruct("locus" -> tv("T"), "alleles" -> TArray(TString()))) { (r, lOff, aOff) =>
-      val returnTuple = r.mb.newLocal[(Locus, IndexedSeq[String])]
-      val locus = getLocus(r, lOff, "T")
-      val alleles = Code.checkcast[IndexedSeq[String]](wrapArg(r, TArray(TString()))(aOff).asInstanceOf[Code[AnyRef]])
-      val tuple = Code.invokeScalaObject[Locus, IndexedSeq[String], (Locus, IndexedSeq[String])](VariantMethods.getClass, "minRep", locus, alleles)
+    registerCode("min_rep", tv("T", "locus"), TArray(TString()), TStruct("locus" -> tv("T"), "alleles" -> TArray(TString()))) {
+      case (r, (locusT: PLocus, lOff), (allelesT, aOff)) =>
+        val returnTuple = r.mb.newLocal[(Locus, IndexedSeq[String])]
+        val locus = getLocus(r, lOff, locusT)
+        val alleles = Code.checkcast[IndexedSeq[String]](wrapArg(r, allelesT)(aOff).asInstanceOf[Code[AnyRef]])
+        val tuple = Code.invokeScalaObject[Locus, IndexedSeq[String], (Locus, IndexedSeq[String])](VariantMethods.getClass, "minRep", locus, alleles)
 
-      val newLocus = Code.checkcast[Locus](returnTuple.load().get[java.lang.Object]("_1"))
-      val newAlleles = Code.checkcast[IndexedSeq[String]](returnTuple.load().get[java.lang.Object]("_2"))
+        val newLocus = Code.checkcast[Locus](returnTuple.load().get[java.lang.Object]("_1"))
+        val newAlleles = Code.checkcast[IndexedSeq[String]](returnTuple.load().get[java.lang.Object]("_2"))
 
-      val srvb = new StagedRegionValueBuilder(r, PTuple(FastIndexedSeq(tv("T").t.physicalType, PArray(PString()))))
-      Code(
-        returnTuple := tuple,
-        srvb.start(),
-        srvb.addBaseStruct(types.coerce[PBaseStruct](tv("T").t.fundamentalType.physicalType), { locusBuilder =>
-          Code(
-            locusBuilder.start(),
-            locusBuilder.addString(newLocus.invoke[String]("contig")),
-            locusBuilder.advance(),
-            locusBuilder.addInt(newLocus.invoke[Int]("position")))
-        }),
-        srvb.advance(),
-        srvb.addArray(PArray(PString()), { allelesBuilder =>
-          Code(
-            allelesBuilder.start(newAlleles.invoke[Int]("size")),
-            Code.whileLoop(allelesBuilder.arrayIdx < newAlleles.invoke[Int]("size"),
-              allelesBuilder.addString(Code.checkcast[String](newAlleles.invoke[Int, java.lang.Object]("apply", allelesBuilder.arrayIdx))),
-              allelesBuilder.advance()))
-        }),
-        srvb.offset)
+        val newLocusT = PLocus(locusT.rg)
+        val newAllelesT = PArray(PString())
+        val srvb = new StagedRegionValueBuilder(r, PTuple(FastIndexedSeq(newLocusT, newAllelesT)))
+        Code(
+          returnTuple := tuple,
+          srvb.start(),
+          srvb.addBaseStruct(newLocusT.fundamentalType.asInstanceOf[PStruct], { locusBuilder =>
```

Comment:
Again outside scope of this PR, wondering if it wouldn't be better to use newLocusT.representation.fundamentalType, not so much to avoid the cast, but because we concretely know the type, which is already PStruct.

Reply:
dunno, seems about the same to me, I think. Probably better your way but I don't think I'll bother to change and repush for this one change.